File: ios/Orch8Example/Orch8Example/Engine/Orch8Manager.swift
import SwiftUI
import UIKit
import Orch8Mobile

struct BannerInfo: Identifiable {
    let id = UUID()
    let title: String
    let message: String
    let style: BannerStyle
}

enum BannerStyle {
    case success, error, warning, info

    var color: Color {
        switch self {
        case .success: return .green
        case .error: return .red
        case .warning: return .orange
        case .info: return .blue
        }
    }

    var icon: String {
        switch self {
        case .success: return "checkmark.circle.fill"
        case .error: return "xmark.circle.fill"
        case .warning: return "exclamationmark.triangle.fill"
        case .info: return "info.circle.fill"
        }
    }
}

struct ApprovalRequest: Identifiable {
    let id = UUID()
    let instanceId: String
    let stepName: String
    let prompt: String
    let choices: [(label: String, value: String)]
}

struct WorkflowStatus: Identifiable {
    let id: String
    let name: String
    var state: String
    var currentStep: String?
    var updatedAt: String
    var dedupKey: String?
    var createdAt: String?
}

@MainActor
class Orch8Manager: ObservableObject {
    @Published var activeBanner: BannerInfo?
    @Published var pendingApproval: ApprovalRequest?
    @Published var activeWorkflows: [WorkflowStatus] = []
    @Published var completedWorkflows: [WorkflowStatus] = []
    @Published var engineReady = false
    @Published var loadedSequenceNames: [String] = []
    @Published var engineInfo: String = ""

    private var engine: MobileEngine?
    private let listener = ExampleEngineListener()
    private var batteryObservers: [NSObjectProtocol] = []
    private var setupTask: Task<Void, Never>?

    static let sequencesUrl = "http://localhost:8080/sequences.json"

    static let workflowNames = [
        "onboarding-flow",
        "payment-verification",
        "feature-access",
    ]

    init() {
        setupTask = Task { await setupEngine() }
    }

    private func setupEngine() async {
        let dbPath = FileManager.default
            .urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("orch8-example.db")
            .path

        let config = MobileEngineConfig(
            tickIntervalMs: 200,
            maxConcurrentSteps: 4,
            maxStepsPerInstance: 1000,
            maxConcurrentInstances: 10,
            maxTickDurationMs: 5000,
            maxInstanceLifetimeSecs: 86400,
            maxStoredSequences: 50,
            maxSequenceSizeBytes: 1_048_576,
            handlerTimeoutMs: 30000,
            operationTimeoutMs: 10000,
            telemetryEnabled: true,
            environment: "development",
            rootPublicKey: "",
            sdkVersion: "0.1.0",
            memoryBudgetBytes: 0,
            sequencesUrl: Self.sequencesUrl
        )

        engineInfo = "tick=\(config.tickIntervalMs)ms | concurrent=\(config.maxConcurrentSteps) steps, \(config.maxConcurrentInstances) instances | timeout=\(config.handlerTimeoutMs / 1000)s | env=\(config.environment)"

        do {
            let eng = try MobileEngine(dbPath: dbPath, config: config)
            registerHandlers(engine: eng)

            listener.onPendingStep = { [weak self] instanceId, stepName, handler in
                Task { @MainActor in
                    self?.handlePendingStep(instanceId: instanceId, stepName: stepName, handler: handler)
                }
            }
            listener.onCompleted = { [weak self] instanceId, output in
                Task { @MainActor in
                    self?.handleCompleted(instanceId: instanceId, output: output)
                }
            }
            listener.onFailed = { [weak self] instanceId, error in
                Task { @MainActor in
                    self?.handleFailed(instanceId: instanceId, error: error)
                }
            }
            eng.setListener(listener: listener)

            let device = UIDevice.current
            let context = DeviceContext(
                deviceId: device.identifierForVendor?.uuidString ?? "unknown",
                osName: device.systemName,
                osVersion: device.systemVersion,
                appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0",
                sdkVersion: "0.1.0"
            )
            try eng.setDeviceContext(context: context)
            fputs("[app] Device context set: \(context.osName) \(context.osVersion), app \(context.appVersion)\n", stderr)

            setupBatteryMonitoring(engine: eng)

            do {
                // Empty string tells the engine to use config.sequencesUrl
                let count = try eng.loadSequencesFromUrl(url: "")
                fputs("[app] Loaded \(count) sequences from \(Self.sequencesUrl)\n", stderr)
            } catch {
                fputs("[app] Failed to load sequences from URL: \(error)\n", stderr)
            }

            refreshLoadedSequences(engine: eng)

            self.engine = eng
            self.engineReady = true

            eng.resume()
        } catch {
            fputs("[app] Failed to initialize engine: \(error)\n", stderr)
        }
    }

    private func setupBatteryMonitoring(engine: MobileEngine) {
        UIDevice.current.isBatteryMonitoringEnabled = true
        reportCurrentBatteryState(engine: engine)

        let stateObserver = NotificationCenter.default.addObserver(
            forName: UIDevice.batteryStateDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let engine = self?.engine else { return }
            self?.reportCurrentBatteryState(engine: engine)
        }

        let levelObserver = NotificationCenter.default.addObserver(
            forName: UIDevice.batteryLevelDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let engine = self?.engine else { return }
            self?.reportCurrentBatteryState(engine: engine)
        }

        batteryObservers = [stateObserver, levelObserver]
    }

    private func reportCurrentBatteryState(engine: MobileEngine) {
        let device = UIDevice.current
        let level = device.batteryLevel
        let uiState = device.batteryState

        let powerState: PowerState
        switch uiState {
        case .charging, .full:
            powerState = .charging
        case .unplugged:
            if level >= 0 && level <= 0.05 {
                powerState = .criticalBattery
            } else if level > 0.05 && level <= 0.20 {
                powerState = .lowBattery
            } else {
                powerState = .unplugged
            }
        default:
            powerState = .unplugged
        }

        do {
            try engine.reportPowerState(state: powerState)
            fputs("[app] Reported power state: \(powerState), battery level: \(level)\n", stderr)
        } catch {
            fputs("[app] Failed to report power state: \(error)\n", stderr)
        }
    }

    private func refreshLoadedSequences(engine: MobileEngine) {
        do {
            let sequences = try engine.loadedSequences()
            loadedSequenceNames = sequences.map { $0.name }
            fputs("[app] Loaded sequences: \(loadedSequenceNames)\n", stderr)
        } catch {
            fputs("[app] Failed to get loaded sequences: \(error)\n", stderr)
        }
    }

    private func registerHandlers(engine: MobileEngine) {
        let handlers: [(String, ExampleStepHandler)] = [
            ("init_profile", ExampleStepHandler { _, input in
                "{\"status\":\"profile_initialized\",\"timestamp\":\"\(ISO8601DateFormatter().string(from: Date()))\"}"
            }),
            ("validate_email", ExampleStepHandler { _, input in
                "{\"valid\":true,\"email\":\"user@example.com\"}"
            }),
            ("show_terms", ExampleStepHandler { _, _ in "{\"shown\":true}" }),
            ("collect_preferences", ExampleStepHandler { _, _ in "{\"collected\":true}" }),
            ("setup_notifications", ExampleStepHandler { _, _ in "{\"configured\":true}" }),
            ("request_approval", ExampleStepHandler { _, _ in "{\"requested\":true}" }),
            ("complete_onboarding", ExampleStepHandler { _, _ in "{\"completed\":true}" }),
            ("init_payment", ExampleStepHandler { _, input in
                "{\"payment_id\":\"PAY-\(UUID().uuidString.prefix(8))\",\"status\":\"initialized\"}"
            }),
            ("validate_amount", ExampleStepHandler { _, _ in "{\"valid\":true,\"amount\":99.99}" }),
            ("fraud_check", ExampleStepHandler { _, _ in "{\"passed\":true,\"score\":0.15}" }),
            ("risk_assessment", ExampleStepHandler { _, _ in "{\"risk_level\":\"low\",\"score\":0.2}" }),
            ("compliance_check", ExampleStepHandler { _, _ in "{\"compliant\":true,\"checks\":[\"AML\",\"KYC\"]}" }),
            ("process_payment", ExampleStepHandler { _, _ in "{\"processed\":true,\"transaction_id\":\"TXN-001\"}" }),
            ("send_receipt", ExampleStepHandler { _, _ in "{\"receipt_sent\":true}" }),
            ("check_eligibility", ExampleStepHandler { _, _ in "{\"eligible\":true}" }),
            ("fetch_feature_config", ExampleStepHandler { _, _ in "{\"feature\":\"premium_analytics\",\"available\":true}" }),
            ("evaluate_rules", ExampleStepHandler { _, _ in "{\"rules_passed\":true}" }),
            ("request_consent", ExampleStepHandler { _, _ in "{\"consent_shown\":true}" }),
            ("verify_identity", ExampleStepHandler { _, _ in "{\"verified\":true,\"method\":\"biometric\"}" }),
            ("activate_feature", ExampleStepHandler { _, _ in "{\"activated\":true,\"feature\":\"premium_analytics\"}" }),
            ("show_banner", ExampleStepHandler { [weak self] _, input in
                if let data = input.data(using: .utf8),
                   let params = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    let title = params["title"] as? String ?? "Notice"
                    let message = params["message"] as? String ?? ""
                    let styleStr = params["style"] as? String ?? "info"
                    let style: BannerStyle = switch styleStr {
                    case "success": .success
                    case "error": .error
                    case "warning": .warning
                    default: .info
                    }
                    Task { @MainActor in
                        self?.activeBanner = BannerInfo(title: title, message: message, style: style)
                    }
                }
                return "{\"banner_shown\":true}"
            }),
        ]

        for (name, handler) in handlers {
            do {
                try engine.registerHandler(name: name, handler: handler)
            } catch {
                fputs("[app] Failed to register handler '\(name)': \(error)\n", stderr)
            }
        }
    }

    func startWorkflow(name: String, input: [String: Any] = [:], dedupKey: String? = nil) {
        guard let engine = engine else { return }
        let inputJson: String
        do {
            let inputData = try JSONSerialization.data(withJSONObject: input)
            inputJson = String(data: inputData, encoding: .utf8) ?? "{}"
        } catch {
            activeBanner = BannerInfo(
                title: "Error",
                message: "Failed to serialize input: \(error.localizedDescription)",
                style: .error
            )
            return
        }

        let generatedDedupKey = dedupKey ?? "\(name)-\(UUID().uuidString.prefix(8))"
        let now = ISO8601DateFormatter().string(from: Date())

        Task.detached { [weak self] in
            do {
                let instanceId = try engine.start(
                    sequenceName: name,
                    input: inputJson,
                    dedupKey: generatedDedupKey
                )
                let status = WorkflowStatus(
                    id: instanceId,
                    name: name,
                    state: "Scheduled",
                    currentStep: nil,
                    updatedAt: now,
                    dedupKey: generatedDedupKey,
                    createdAt: now
                )
                await MainActor.run {
                    self?.activeWorkflows.append(status)
                }
                fputs("[app] Started workflow '\(name)' instance=\(instanceId) dedup=\(generatedDedupKey)\n", stderr)
            } catch {
                fputs("[app] Failed to start workflow '\(name)': \(error)\n", stderr)
                await MainActor.run {
                    self?.activeBanner = BannerInfo(
                        title: "Error",
                        message: "Failed to start \(name): \(error.localizedDescription)",
                        style: .error
                    )
                }
            }
        }
    }

    func dismissBanner() {
        activeBanner = nil
    }

    func resolveApproval(decision: String) {
        guard let approval = pendingApproval, let engine = engine else { return }
        pendingApproval = nil

        if let idx = activeWorkflows.firstIndex(where: { $0.id == approval.instanceId }) {
            activeWorkflows[idx].state = "Running"
            activeWorkflows[idx].currentStep = "Processing \(decision)..."
        }

        Task.detached {
            do {
                let outputData = try JSONSerialization.data(withJSONObject: ["value": decision])
                let output = String(data: outputData, encoding: .utf8) ?? "{}"
                try engine.completeStep(
                    instanceId: approval.instanceId,
                    stepName: approval.stepName,
                    output: output
                )
            } catch {
                fputs("[app] Failed to resolve approval: \(error)\n", stderr)
            }
        }
    }

    func refreshWorkflowStates() {
        guard let engine = engine else { return }
        do {
            let active = try engine.activeInstances()
            for summary in active {
                if let idx = activeWorkflows.firstIndex(where: { $0.id == summary.instanceId }) {
                    activeWorkflows[idx].state = "\(summary.state)"
                }
            }
        } catch {
            fputs("[app] Failed to refresh states: \(error)\n", stderr)
        }
    }

    func cancelWorkflow(instanceId: String) {
        guard let engine = engine else { return }
        activeWorkflows.removeAll { $0.id == instanceId }

        Task.detached {
            do {
                try engine.cancelInstance(instanceId: instanceId)
            } catch {
                fputs("[app] Failed to cancel workflow: \(error)\n", stderr)
            }
        }
    }

    func pauseEngine() {
        guard let engine = engine else { return }
        do {
            try engine.pause()
            fputs("[app] Engine paused\n", stderr)
        } catch {
            fputs("[app] Failed to pause engine: \(error)\n", stderr)
        }
    }

    func resumeEngine() {
        guard let engine = engine else { return }
        engine.resume()
        fputs("[app] Engine resumed\n", stderr)
    }

    func shutdownEngine() {
        setupTask?.cancel()
        setupTask = nil

        guard let engine = engine else { return }
        do {
            try engine.flushTelemetry()
            try engine.shutdown()
            fputs("[app] Engine shut down\n", stderr)
        } catch {
            fputs("[app] Failed to shut down engine: \(error)\n", stderr)
        }

        for observer in batteryObservers {
            NotificationCenter.default.removeObserver(observer)
        }
        batteryObservers.removeAll()
    }

    func reportPowerState(_ state: PowerState) {
        guard let engine = engine else { return }
        do {
            try engine.reportPowerState(state: state)
            fputs("[app] Manually reported power state: \(state)\n", stderr)
        } catch {
            fputs("[app] Failed to report power state: \(error)\n", stderr)
        }
    }

    func onPushReceived() {
        guard let engine = engine else { return }
        do {
            try engine.onPushReceived()
            fputs("[app] Push received — forced immediate sync\n", stderr)
        } catch {
            fputs("[app] Failed to handle push: \(error)\n", stderr)
        }
    }

    func getLoadedSequences() -> [SequenceInfo] {
        guard let engine = engine else { return [] }
        do {
            return try engine.loadedSequences()
        } catch {
            fputs("[app] Failed to get loaded sequences: \(error)\n", stderr)
            return []
        }
    }

    func getInstanceDetail(instanceId: String) -> InstanceSummary? {
        guard let engine = engine else { return nil }
        do {
            return try engine.getInstance(instanceId: instanceId)
        } catch {
            fputs("[app] Failed to get instance detail: \(error)\n", stderr)
            return nil
        }
    }

    private func handlePendingStep(instanceId: String, stepName: String, handler: String) {
        if let idx = activeWorkflows.firstIndex(where: { $0.id == instanceId }) {
            activeWorkflows[idx].state = "Waiting"
            activeWorkflows[idx].currentStep = stepName
        }

        if handler == "request_approval" || handler == "show_terms" || handler == "request_consent" {
            let prompt: String
            let choices: [(String, String)]
            switch handler {
            case "show_terms":
                prompt = "Please review and accept the Terms of Service."
                choices = [("Accept", "accepted"), ("Decline", "declined")]
            case "request_consent":
                prompt = "This feature collects usage data. Do you consent?"
                choices = [("I Agree", "consented"), ("No Thanks", "declined")]
            default:
                prompt = "Action requires approval. Please review and decide."
                choices = [("Approve", "approved"), ("Reject", "rejected")]
            }
            pendingApproval = ApprovalRequest(
                instanceId: instanceId,
                stepName: stepName,
                prompt: prompt,
                choices: choices
            )
        } else if handler == "collect_preferences" || handler == "setup_notifications" {
            guard let engine = engine else { return }
            let output: String
            switch handler {
            case "collect_preferences":
                output = "{\"value\":\"tech,news\"}"
            default:
                output = "{\"value\":\"enabled\"}"
            }
            Task.detached {
                do {
                    try engine.completeStep(instanceId: instanceId, stepName: stepName, output: output)
                } catch {
                    fputs("[app] Failed to auto-complete step \(stepName): \(error)\n", stderr)
                }
            }
        }
    }

    private func handleCompleted(instanceId: String, output: String) {
        if let idx = activeWorkflows.firstIndex(where: { $0.id == instanceId }) {
            var workflow = activeWorkflows.remove(at: idx)
            workflow.state = "Completed"
            workflow.updatedAt = ISO8601DateFormatter().string(from: Date())
            completedWorkflows.insert(workflow, at: 0)
        }
    }

    private func handleFailed(instanceId: String, error: String) {
        if let idx = activeWorkflows.firstIndex(where: { $0.id == instanceId }) {
            var workflow = activeWorkflows.remove(at: idx)
            workflow.state = "Failed"
            workflow.updatedAt = ISO8601DateFormatter().string(from: Date())
            completedWorkflows.insert(workflow, at: 0)
        }
        activeBanner = BannerInfo(
            title: "Workflow Failed",
            message: error,
            style: .error
        )
    }
}

File: android/app/src/main/java/io/orch8/example/Orch8Manager.kt
package io.orch8.example

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import io.orch8.mobile.DeviceContext
import io.orch8.mobile.EngineListener
import io.orch8.mobile.InstanceSummary
import io.orch8.mobile.MobileEngine
import io.orch8.mobile.MobileEngineConfig
import io.orch8.mobile.PowerState
import io.orch8.mobile.SequenceInfo
import io.orch8.mobile.StepHandler
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

data class BannerInfo(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val message: String,
    val style: BannerStyle
)

enum class BannerStyle {
    SUCCESS, ERROR, WARNING, INFO
}

data class ApprovalRequest(
    val id: String = UUID.randomUUID().toString(),
    val instanceId: String,
    val stepName: String,
    val prompt: String,
    val choices: List<Pair<String, String>>
)

data class WorkflowStatus(
    val id: String,
    val name: String,
    var state: String,
    var currentStep: String? = null,
    var updatedAt: String = "",
    val dedupKey: String? = null,
    val createdAt: String? = null
)

data class WorkflowDefinition(
    val name: String,
    val displayName: String,
    val description: String,
)

class Orch8Manager(private val context: Context) {
    companion object {
        private const val SEQUENCES_URL = "http://10.0.2.2:8080/sequences.json"
        private const val PREFS_NAME = "orch8_prefs"
        private const val KEY_DEVICE_ID = "device_id"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var engine: MobileEngine? = null

    private val _activeBanner = MutableStateFlow<BannerInfo?>(null)
    val activeBanner: StateFlow<BannerInfo?> = _activeBanner.asStateFlow()

    private val _pendingApproval = MutableStateFlow<ApprovalRequest?>(null)
    val pendingApproval: StateFlow<ApprovalRequest?> = _pendingApproval.asStateFlow()

    private val _activeWorkflows = MutableStateFlow<List<WorkflowStatus>>(emptyList())
    val activeWorkflows: StateFlow<List<WorkflowStatus>> = _activeWorkflows.asStateFlow()

    private val _completedWorkflows = MutableStateFlow<List<WorkflowStatus>>(emptyList())
    val completedWorkflows: StateFlow<List<WorkflowStatus>> = _completedWorkflows.asStateFlow()

    private val _engineReady = MutableStateFlow(false)
    val engineReady: StateFlow<Boolean> = _engineReady.asStateFlow()

    private val _loadedSequenceNames = MutableStateFlow<List<String>>(emptyList())
    val loadedSequenceNames: StateFlow<List<String>> = _loadedSequenceNames.asStateFlow()

    private val _engineInfo = MutableStateFlow("")
    val engineInfo: StateFlow<String> = _engineInfo.asStateFlow()

    val workflows = listOf(
        WorkflowDefinition(
            "onboarding-flow", "User Onboarding",
            "10 steps — email validation, terms acceptance (wait), preferences (wait), notifications (wait), admin approval (wait), conditional routing, success/rejection banners",
        ),
        WorkflowDefinition(
            "payment-verification", "Payment Verification",
            "10 steps — amount validation, fraud check, risk assessment, compliance, payment approval (wait), conditional routing, success banner + receipt or rejection banner",
        ),
        WorkflowDefinition(
            "feature-access", "Feature Access Control",
            "10 steps — eligibility, config, rules, consent (wait), identity verification, access approval (wait), dual routing, access granted or denied/limited banners",
        ),
    )

    private val batteryReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, intent: Intent?) {
            if (intent?.action != Intent.ACTION_BATTERY_CHANGED) return
            val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
            val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
            if (level < 0 || scale <= 0) return
            val pct = (level * 100) / scale

            val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
            val isCharging = status == BatteryManager.BATTERY_STATUS_CHARGING
                    || status == BatteryManager.BATTERY_STATUS_FULL

            val powerState = when {
                isCharging -> PowerState.CHARGING
                pct > 20 -> PowerState.UNPLUGGED
                pct >= 5 -> PowerState.LOW_BATTERY
                else -> PowerState.CRITICAL_BATTERY
            }
            reportPowerState(powerState)
        }
    }

    init {
        scope.launch(Dispatchers.IO) { setupEngine() }
    }

    private fun getOrCreateDeviceId(): String {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        var deviceId = prefs.getString(KEY_DEVICE_ID, null)
        if (deviceId == null) {
            deviceId = "android-${UUID.randomUUID()}"
            prefs.edit().putString(KEY_DEVICE_ID, deviceId).apply()
        }
        return deviceId
    }

    private fun setupEngine() {
        val dbPath = context.getDatabasePath("orch8-example.db").absolutePath

        val tickIntervalMs: ULong = 200u
        val maxConcurrentSteps: UInt = 4u
        val maxConcurrentInstances: UInt = 10u

        val config = MobileEngineConfig(
            tickIntervalMs = tickIntervalMs,
            maxConcurrentSteps = maxConcurrentSteps,
            maxStepsPerInstance = 1000u,
            maxConcurrentInstances = maxConcurrentInstances,
            maxTickDurationMs = 5000u,
            maxInstanceLifetimeSecs = 86400u,
            maxStoredSequences = 50u,
            maxSequenceSizeBytes = 1_048_576u,
            handlerTimeoutMs = 30000u,
            operationTimeoutMs = 10000u,
            telemetryEnabled = true,
            environment = "development",
            rootPublicKey = "",
            sdkVersion = "0.1.0",
            memoryBudgetBytes = 0u,
            sequencesUrl = SEQUENCES_URL
        )

        try {
            val eng = MobileEngine(dbPath, config)
            registerHandlers(eng)
            eng.setListener(ExampleListener())

            try {
                val count = eng.loadSequencesFromUrl("")
                android.util.Log.i("Orch8Manager", "Loaded $count sequences from $SEQUENCES_URL")
            } catch (e: Exception) {
                android.util.Log.e("Orch8Manager", "Failed to load sequences from URL", e)
            }

            eng.setDeviceContext(DeviceContext(
                deviceId = getOrCreateDeviceId(),
                osName = "Android",
                osVersion = android.os.Build.VERSION.RELEASE,
                appVersion = "1.0.0",
                sdkVersion = "0.1.0"
            ))

            engine = eng

            scope.launch(Dispatchers.Main) {
                _engineReady.value = true
                _engineInfo.value = "tick=${tickIntervalMs}ms | workers=$maxConcurrentSteps | maxInstances=$maxConcurrentInstances"
            }

            refreshLoadedSequences()
            eng.resume()
        } catch (e: Exception) {
            android.util.Log.e("Orch8Manager", "Failed to initialize engine", e)
        }
    }

    private fun refreshLoadedSequences() {
        val eng = engine ?: return
        try {
            val sequences = eng.loadedSequences()
            scope.launch(Dispatchers.Main) {
                _loadedSequenceNames.value = sequences.map { "${it.name} v${it.version}" }
            }
        } catch (e: Exception) {
            android.util.Log.e("Orch8Manager", "Failed to list loaded sequences", e)
        }
    }

    private fun registerHandlers(engine: MobileEngine) {
        val handlers = mapOf<String, (String, String) -> String>(
            "init_profile" to { _, _ ->
                val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())
                """{"status":"profile_initialized","timestamp":"$timestamp"}"""
            },
            "validate_email" to { _, _ -> """{"valid":true,"email":"user@example.com"}""" },
            "show_terms" to { _, _ -> """{"shown":true}""" },
            "collect_preferences" to { _, _ -> """{"collected":true}""" },
            "setup_notifications" to { _, _ -> """{"configured":true}""" },
            "request_approval" to { _, _ -> """{"requested":true}""" },
            "complete_onboarding" to { _, _ -> """{"completed":true}""" },
            "init_payment" to { _, _ -> """{"payment_id":"PAY-${UUID.randomUUID().toString().take(8)}","status":"initialized"}""" },
            "validate_amount" to { _, _ -> """{"valid":true,"amount":99.99}""" },
            "fraud_check" to { _, _ -> """{"passed":true,"score":0.15}""" },
            "risk_assessment" to { _, _ -> """{"risk_level":"low","score":0.2}""" },
            "compliance_check" to { _, _ -> """{"compliant":true,"checks":["AML","KYC"]}""" },
            "process_payment" to { _, _ -> """{"processed":true,"transaction_id":"TXN-001"}""" },
            "send_receipt" to { _, _ -> """{"receipt_sent":true}""" },
            "check_eligibility" to { _, _ -> """{"eligible":true}""" },
            "fetch_feature_config" to { _, _ -> """{"feature":"premium_analytics","available":true}""" },
            "evaluate_rules" to { _, _ -> """{"rules_passed":true}""" },
            "request_consent" to { _, _ -> """{"consent_shown":true}""" },
            "verify_identity" to { _, _ -> """{"verified":true,"method":"biometric"}""" },
            "activate_feature" to { _, _ -> """{"activated":true,"feature":"premium_analytics"}""" },
            "show_banner" to { _, input ->
                try {
                    val params = JSONObject(input)
                    val title = params.optString("title", "Notice")
                    val message = params.optString("message", "")
                    val styleStr = params.optString("style", "info")
                    val style = when (styleStr) {
                        "success" -> BannerStyle.SUCCESS
                        "error" -> BannerStyle.ERROR
                        "warning" -> BannerStyle.WARNING
                        else -> BannerStyle.INFO
                    }
                    scope.launch(Dispatchers.Main) {
                        _activeBanner.value = BannerInfo(title = title, message = message, style = style)
                    }
                } catch (_: Exception) {}
                """{"banner_shown":true}"""
            }
        )

        for ((name, handler) in handlers) {
            engine.registerHandler(name, object : StepHandler {
                override fun execute(stepName: String, input: String): String {
                    return handler(stepName, input)
                }
            })
        }
    }

    fun startWorkflow(name: String, input: Map<String, Any> = emptyMap()) {
        val eng = engine ?: return
        val inputJson = JSONObject(input).toString()
        val dedupKey = "$name-${UUID.randomUUID().toString().take(8)}"
        scope.launch(Dispatchers.IO) {
            try {
                val instanceId = eng.start(name, inputJson, dedupKey)
                val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())
                val status = WorkflowStatus(
                    id = instanceId,
                    name = name,
                    state = "Scheduled",
                    updatedAt = timestamp,
                    dedupKey = dedupKey,
                    createdAt = timestamp
                )
                scope.launch(Dispatchers.Main) {
                    _activeWorkflows.value = _activeWorkflows.value + status
                }
                android.util.Log.i("Orch8Manager", "Started workflow '$name' instance=$instanceId dedup=$dedupKey")
            } catch (e: Exception) {
                android.util.Log.e("Orch8Manager", "Failed to start workflow '$name'", e)
                scope.launch(Dispatchers.Main) {
                    _activeBanner.value = BannerInfo(
                        title = "Error",
                        message = "Failed to start $name: ${e.message}",
                        style = BannerStyle.ERROR
                    )
                }
            }
        }
    }

    fun dismissBanner() {
        _activeBanner.value = null
    }

    fun resolveApproval(decision: String) {
        val approval = _pendingApproval.value ?: return
        val eng = engine ?: return
        _pendingApproval.value = null

        _activeWorkflows.value = _activeWorkflows.value.map { wf ->
            if (wf.id == approval.instanceId) wf.copy(state = "Running", currentStep = "Processing $decision...")
            else wf
        }

        scope.launch(Dispatchers.IO) {
            try {
                val output = JSONObject().put("value", decision).toString()
                eng.completeStep(approval.instanceId, approval.stepName, output)
            } catch (e: Exception) {
                android.util.Log.e("Orch8Manager", "Failed to resolve approval", e)
            }
        }
    }

    fun cancelWorkflow(instanceId: String) {
        val eng = engine ?: return
        _activeWorkflows.value = _activeWorkflows.value.filter { it.id != instanceId }

        scope.launch(Dispatchers.IO) {
            try {
                eng.cancelInstance(instanceId)
            } catch (e: Exception) {
                android.util.Log.e("Orch8Manager", "Failed to cancel workflow", e)
            }
        }
    }

    fun resumeEngine() {
        engine?.resume()
        android.util.Log.i("Orch8Manager", "Engine resumed")
    }

    fun pauseEngine() {
        engine?.pause()
        android.util.Log.i("Orch8Manager", "Engine paused")
    }

    fun shutdownEngine() {
        engine?.let { eng ->
            eng.flushTelemetry()
            eng.shutdown()
        }
        engine = null
        android.util.Log.i("Orch8Manager", "Engine shut down")
    }

    fun destroy() {
        unregisterBatteryReceiver()
        shutdownEngine()
        scope.cancel()
    }

    fun reportPowerState(state: PowerState) {
        engine?.reportPowerState(state)
        android.util.Log.d("Orch8Manager", "Reported power state: $state")
    }

    fun onPushReceived() {
        engine?.onPushReceived()
        android.util.Log.i("Orch8Manager", "Push received — forced immediate sync")
    }

    fun getLoadedSequences(): List<SequenceInfo> {
        return engine?.loadedSequences() ?: emptyList()
    }

    fun getInstanceDetail(instanceId: String): InstanceSummary? {
        return try {
            engine?.getInstance(instanceId)
        } catch (e: Exception) {
            android.util.Log.e("Orch8Manager", "Failed to get instance detail", e)
            null
        }
    }

    fun registerBatteryReceiver() {
        val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        context.registerReceiver(batteryReceiver, filter)
        android.util.Log.i("Orch8Manager", "Battery receiver registered")
    }

    fun unregisterBatteryReceiver() {
        try {
            context.unregisterReceiver(batteryReceiver)
        } catch (_: IllegalArgumentException) {
            // Receiver was not registered
        }
    }

    private inner class ExampleListener : EngineListener {
        override fun onInstanceCompleted(instanceId: String, output: String) {
            scope.launch(Dispatchers.Main) {
                val workflow = _activeWorkflows.value.find { it.id == instanceId }
                if (workflow != null) {
                    _activeWorkflows.value = _activeWorkflows.value.filter { it.id != instanceId }
                    val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())
                    _completedWorkflows.value = listOf(
                        workflow.copy(state = "Completed", updatedAt = timestamp)
                    ) + _completedWorkflows.value
                }
            }
        }

        override fun onInstanceFailed(instanceId: String, error: String) {
            scope.launch(Dispatchers.Main) {
                val workflow = _activeWorkflows.value.find { it.id == instanceId }
                if (workflow != null) {
                    _activeWorkflows.value = _activeWorkflows.value.filter { it.id != instanceId }
                    val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())
                    _completedWorkflows.value = listOf(
                        workflow.copy(state = "Failed", updatedAt = timestamp)
                    ) + _completedWorkflows.value
                }
                _activeBanner.value = BannerInfo(
                    title = "Workflow Failed",
                    message = error,
                    style = BannerStyle.ERROR
                )
            }
        }

        override fun onStepPending(instanceId: String, stepName: String, handler: String) {
            scope.launch(Dispatchers.Main) {
                _activeWorkflows.value = _activeWorkflows.value.map { wf ->
                    if (wf.id == instanceId) wf.copy(state = "Waiting", currentStep = stepName)
                    else wf
                }

                when (handler) {
                    "request_approval" -> {
                        _pendingApproval.value = ApprovalRequest(
                            instanceId = instanceId,
                            stepName = stepName,
                            prompt = "Action requires approval. Please review and decide.",
                            choices = listOf("Approve" to "approved", "Reject" to "rejected")
                        )
                    }
                    "show_terms" -> {
                        _pendingApproval.value = ApprovalRequest(
                            instanceId = instanceId,
                            stepName = stepName,
                            prompt = "Please review and accept the Terms of Service.",
                            choices = listOf("Accept" to "accepted", "Decline" to "declined")
                        )
                    }
                    "request_consent" -> {
                        _pendingApproval.value = ApprovalRequest(
                            instanceId = instanceId,
                            stepName = stepName,
                            prompt = "This feature collects usage data. Do you consent?",
                            choices = listOf("I Agree" to "consented", "No Thanks" to "declined")
                        )
                    }
                    "collect_preferences", "setup_notifications" -> {
                        val eng = engine ?: return@launch
                        val output = when (handler) {
                            "collect_preferences" -> """{"value":"tech,news"}"""
                            else -> """{"value":"enabled"}"""
                        }
                        scope.launch(Dispatchers.IO) {
                            try {
                                eng.completeStep(instanceId, stepName, output)
                            } catch (e: Exception) {
                                android.util.Log.e("Orch8Manager", "Failed to auto-complete step $stepName", e)
                            }
                        }
                    }
                }
            }
        }
    }
}
